File: LeetCode/src/array/ProductOfArrayExceptSelf.java
package array;

public class ProductOfArrayExceptSelf {
    public int[] productExceptSelf(int[] nums) {
        int n = nums.length;
        int[] results = new int[n];

        int product = 1;

        for (int i = 0; i < n; i++) {
            results[i] = product;
            product *= nums[i];
        }

        product = 1;
        for (int i = n - 1; i >= 0; i--) {
            results[i] = product * results[i];
            product *= nums[i];
        }

        return results;
    }

    public static void main(String[] args) {
        ProductOfArrayExceptSelf solution = new ProductOfArrayExceptSelf();
        int[] nums = {1, 2, 3, 4};
        int[] result = solution.productExceptSelf(nums);
        System.out.print("[");
        for (int i = 0; i < result.length; i++) {
            System.out.print(result[i]);
            if (i < result.length - 1) {
                System.out.print(", ");
            }
        }
        System.out.println("]");//Output: [24, 12, 8, 6]
    }
}

File: LeetCode_Python/prefixsum/ProductOfArrayExceptSelf.py
class ProductOfArrayExceptSelf:
    def productExceptSelf(self, nums: list[int]) -> list[int]:
        n = len(nums)
        product = 1
        res = [1] * n

        for i in range(n):
            res[i] = product
            product *= nums[i]

        product = 1

        for i in range(n - 1, -1, -1):
            res[i] = product * res[i]
            product *= nums[i]

        return res


if __name__ == "__main__":
    solution = ProductOfArrayExceptSelf()

    print("Expected Output: [24,12,8,6]")
    print("Actual Output:", solution.productExceptSelf([1, 2, 3, 4]))

    print("Expected Output: [0,0,9,0,0]")
    print("Actual Output:", solution.productExceptSelf([-1, 1, 0, -3, 3]))

    print("Expected Output: [0,0,0,0]")
    print("Actual Output:", solution.productExceptSelf([0, 0, 0, 0]))
